appcelerator/titanium_mobile
# [TIMOB-24695] Android: URL auto-encoding improvements

**JIRA:** https://jira.appcelerator.org/browse/TIMOB-24695

**Changes:**
- No longer encodes legal chars `!`, `$`, `'`, `(`, `)`, `+`, `,`, `/`, `:`, `;`, `?`, and `@` in query params.
- No longer encodes legal chars `$`, `&`, `+`, `,`, `;`, and `=` in URL's "username:password" component.
- No longer encodes legal chars `$`, `&`, `+`, `,`, `;`, `=`, `:`, `@`, `/`, and `?` in URL #fragment.
- Spaces in the URL's query paramaters are now encoded as `+` instead of `%20`.
- Given URL's %-encoded characters are now always preserved.
  * Used to decode `%2F` to `/` in URL path, which can break the path.

**Test Procedure:**

1. Run the below code on an Android device that has Internet access.
2. Tap the "URL Space Test" button.
3. An alert dialog will appear. Verify that it displays the following:
`
Received HTTP Request:
POST /Space%20Test?test=Space+Test HTTP/1.1
`
4. Close the alert dialog.
5. Tap the "URL Path %2F Test" button.
6. An alert dialog will appear. Verify that it displays the following:
`
Received HTTP Request:
POST /Slash%2FTest HTTP/1.1
`
7. Close the alert dialog.
8. Tap the "URL Query Encoding Test" button.
9. An alert dialog will appear. Verify that it displays the following:
`
Received HTTP Request:
POST /?test=%22(!$'+,/:;?@)%22 HTTP/1.1
`
10. Close the alert dialog.
11. Tap the "URL Query Slash Test" button.
12. An alert dialog will appear. Verify that it displays the following:
`
Received HTTP Request:
POST /?test=abc/xyz HTTP/1.1
`
13. Close the alert dialog.
14. Verify that an image is displayed at the bottom of the screen. (Was loaded via URL.)

**Test Code:**
```javascript
var listenSocket = Ti.Network.Socket.createTCP(
{
	host : "localhost",
	port : 40404,
	accepted : function(e)
	{
		Ti.Stream.pump(e.inbound, function(e)
		{
			var message = "<empty>";
			if (e.buffer) {
				message = e.buffer.toString();
				var index = message.indexOf("\n");
				if (index > 0) {
					message = message.substr(0, index);
				}
			}
			alert("Received HTTP Request:\n" + message);
			listenSocket.accept({ timeout: 30000 });
		}, 1024, true);
	},
});
listenSocket.listen();
listenSocket.accept({ timeout: 30000 });

function createHttpButton(text, url) {
	var newButton = Ti.UI.createButton(
	{
		title: text,
		top: "5%",
	});
	newButton.addEventListener("click", function(e)
	{
		var httpClient = Ti.Network.createHTTPClient();
		httpClient.open("POST", url);
		httpClient.send();
	});
	return newButton;
}

var window = Titanium.UI.createWindow({ layout: "vertical" });
window.add(createHttpButton("URL Space Test", "http://localhost:40404/Space Test?test=Space Test"));
window.add(createHttpButton("URL Path %2F Test", "http://localhost:40404/Slash%2FTest"));
window.add(createHttpButton("URL Query Encoding Test", "http://localhost:40404/?test=\"(!$'+,/:;?@)\""));
window.add(createHttpButton("URL Query Slash Test", "http://localhost:40404?test=abc/xyz"));
window.add(Ti.UI.createImageView(
{
	image: "http://scontent.xx.fbcdn.net.rsz.io/v/t1.0-9/s720x720/17795851_419371511758070_7481565181564022529_n.jpg?oh=7772aea44d2f40fdbd42285ca61d7aac&oe=59771D60?mode=crop&width=333&height=250",
	top: "5%",
}));
window.open();
```
